### PR TITLE
feat(vectorize): Nx.Defn.Expr.cond

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: nx/lib/nx/defn/expr.ex [hunk 1]
```diff
@@ -137,18 +137,33 @@ defmodule Nx.Defn.Expr do
...
+    if vectorized_axes == [] do
+      out
+    else
+      {result, []} =
+        Composite.traverse(out, vectorized_axes, fn expr, [axes | tail] ->
+          {Nx.vectorize(expr, axes), tail}
+        end)
+
+      result
+    end
```
Comment (josevalim) [inline nx/lib/nx/defn/expr.ex:166]: We may use this in `while`as well, so I can see we encapsulating this soon. :)

Comment (polvalente) [reply nx/lib/nx/defn/expr.ex:166]: yup. we might even be able to absorb it into `flatten_to_composite` somehow

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -4770,20 +4799,30 @@ defmodule Nx do
...
     {devectorized_tensors, canonical, n}
   end
 
-  defp do_reshape_vectors_devectorize(tensors, [], _n), do: tensors
+  defp do_reshape_vectors_devectorize(tensors, [], _n, _make_broadcastable), do: tensors
+
+  defp do_reshape_vectors_devectorize(tensors, canonical_vectorized_axes, n, make_broadcastable) do
+    rank =
+      Enum.reduce(
+        tensors,
```
Comment (josevalim) [inline nx/lib/nx.ex:4821]:
```suggestion
        tl(tensors),
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -4625,6 +4625,13 @@ defmodule Nx do
   For vectors with the same name to be compatible, they need to either
   have the same size or one must be of size 1.
 
+  ## Options
+
+    * `:make_broadcastable` - boolean that indicates whether the inner
+      shapes should also be modified. If `true`, 1-sized leading dimensions
```
Comment (josevalim) [inline nx/lib/nx.ex:4631]: It seems we changed the default to `true`? Also, if the default is `true`, perhaps an option `vectors_only: true` would be better?

Comment (polvalente) [reply nx/lib/nx.ex:4631]: Yes we did, as it is the most common use case and cleans up a lot of the expression graphs.

Comment (polvalente) [reply nx/lib/nx.ex:4631]: Maybe "keep_inner_shape" instead? vectors only might convey that only tensors that are already vectorized would be changed

**Review by josevalim — APPROVED** [review]

File: nx/lib/nx.ex [hunk 4]
```diff
@@ -4749,10 +4772,16 @@ defmodule Nx do
...
   @doc type: :shape
-  def broadcast_vectors([t]), do: [to_tensor(t)]
+  def broadcast_vectors(tensors_or_containers, opts \\ [])
 
-  def broadcast_vectors(tensors) when is_list(tensors) do
-    {devectorized_tensors, target_vectorized_axes, offset} = do_reshape_vectors(tensors)
+  def broadcast_vectors([t], _opts), do: [to_tensor(t)]
+
+  def broadcast_vectors(tensors, opts) when is_list(tensors) do
+    # mode: broadcastable | vectorized_only
```
Comment (polvalente) [inline nx/lib/nx.ex:4780]:
```suggestion
```

File: nx/lib/nx.ex [hunk 5]
```diff
@@ -4749,10 +4774,16 @@ defmodule Nx do
...
```
Comment (polvalente) [inline nx/lib/nx.ex:4782]:
```suggestion
```

**Pull request merged**